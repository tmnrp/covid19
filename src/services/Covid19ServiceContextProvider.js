import React, { useState, useEffect } from "react";
import { fetchJHCSSEData } from "../services/Covid19Service";
import { showLoader, hideLoader } from "../components/loader/Loader";

export const Covid19Context = React.createContext();

const Covid19ServiceContextProvider = (props) => {
  const [records, setRecords] = useState([]);

  useEffect(() => onCmpMount(fetchJHCSSEData, setRecords, callback), []);

  return (
    <Covid19Context.Provider
      data-testid="covid19-context-provider"
      value={{ records: records, setRecords: setRecords }}
    >
      {props.children}
    </Covid19Context.Provider>
  );
};

export const onCmpMount = (fetchJHCSSEData, setRecords, callback) => {
  showLoader();
  fetchJHCSSEData((data) => callback(data, setRecords));
};

export const callback = (data, setRecords) => {
  if (typeof setRecords === "function") {
    setRecords(data);
  }
  hideLoader();
};

export default Covid19ServiceContextProvider;
