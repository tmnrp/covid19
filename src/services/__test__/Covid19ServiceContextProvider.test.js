import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Covid19ServiceContextProvider, {
  onCmpMount,
  callback
} from "../Covid19ServiceContextProvider";

test("Test the callback with valid params", () => {
  const { container } = render(
    <Covid19ServiceContextProvider>
      <p>Testing</p>
    </Covid19ServiceContextProvider>
  );
  expect(container).toMatchSnapshot();
});

test("Test the callback with valid params", () => {
  let record = [];
  const expectedRecord = [1, 2, 3];
  const setRecords = (data) => {
    record = data;
  };

  const { getByTestId } = render(
    <div data-testid="myid" id="myid" className="loader-container is-active">
      ABC
    </div>
  );

  callback(expectedRecord, setRecords);
  expect(getByTestId("myid")).not.toHaveClass("loader-container is-active");
  expect(record).toEqual(expectedRecord);
});

test("Test the callback with invalid typeof for setRecords", () => {
  let record = [];
  const expectedRecord = [1, 2, 3];
  const setRecords = "";

  const { getByTestId } = render(
    <div data-testid="myid" id="myid" className="loader-container is-active">
      ABC
    </div>
  );

  callback(expectedRecord, setRecords);
  expect(getByTestId("myid")).not.toHaveClass("loader-container is-active");
  expect(record).toEqual(record);
});

test("onCmpMount", () => {
  const { getByTestId } = render(
    <div data-testid="loadmask" id="loadmask" className="loader-container">
      ABC
    </div>
  );

  let record = [];
  const expectedRecord = [1, 2, 3];
  const setRecords = (data) => {
    record = data;
  };
  const fetchJHCSSEData = (callback) => {
    callback(expectedRecord, setRecords);
  };

  onCmpMount(fetchJHCSSEData, setRecords, callback);
  expect(getByTestId("loadmask")).toHaveClass("loader-container");
  expect(record).toEqual(expectedRecord);
});
