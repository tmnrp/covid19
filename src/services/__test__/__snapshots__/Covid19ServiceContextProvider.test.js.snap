// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the callback with valid params 1`] = `
<div>
  <p>
    Testing
  </p>
</div>
`;
